rfc/20240824-provider-cache-locking.md

```diff
@@ -0,0 +1,68 @@
+# Safe Global Provider Cache Access
+
+Issue: https://github.com/OpenTofu/opentofu/issues/1483
+
+Many CI/CD systems download all providers for every single tofu action run. The global provider cache allows downloaded providers to be saved in between runs and linked into the local provider cache (in .terraform/). Unfortunately, the global provider cache does not have any locking around it and can fail when used in multiple simultaneous runs.
+
+Another common use case is terragrunt. When running `tofu init` on multiple projects simultaneously via the tool, there is a high likely hood of a conflicting write to a provider's directory. This has caused some interesting tooling workarounds on their end, which are not ideal.
```

Comment:
It might be worth explaining that this is mainly when using terragrunt's `run-all` or `graph` commands.

Reply:
Not entirely accurate. 

Unless users are using [terragrunt-fetch-dependency-output-from-state](https://terragrunt.gruntwork.io/docs/reference/cli-options/#terragrunt-fetch-dependency-output-from-state), users can still end up writing providers to disk for multiple Units (the directories where `terragrunt.hcl` files live) concurrently for dependencies.

Reply:
Also, this problem is not unique to `terragrunt run-all`. The same problem is in atlantis/jenkins with parallel stages, no matter if you use terragrunt or not.